### PR TITLE
Implement n-dimensional hermitian FFTs

### LINKED ISSUE #59127: Hermitian symmetric Fast Fourier Transform
## 🚀 Feature
I would be very helpful if we had an implementation of the HFFTN like we have for the [RFFTN](https://pytorch.org/docs/stable/fft.html#torch.fft.rfftn)

## Motivation

Currently, the [HFFT](https://pytorch.org/docs/stable/fft.html#torch.fft.hfft) only supports a single dimension.

## Pitch

Just like the `torch.fft.rfft`, It is desirable to have the following

- [x] ~`torch.fft.hfft`~ `Already present`
- [x] ~`torch.fft.ihfft`~ `Already present`
- [ ] `torch.fft.hfft2`
- [ ] `torch.fft.ihfft2`
- [ ] `torch.fft.hfftn`
- [ ] `torch.fft.ihfftn`

## Alternatives

I am currently checking [`Aten/native`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SpectralOps.cpp) so see if it is possible to have a `custom C++ extension` for it. But a nice python wrapper as for the `torch.fft.rfft` would be very very helpful.

cc @mruberry @peterbell10 @walterddr

### PR DESCRIPTION
**Pull request by peterbell10**
Closes #59127

cc @mruberry @peterbell10 @walterddr

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -10135,6 +10135,26 @@
   python_module: fft
   variants: function
 
+- func: fft_hfft2(Tensor self, int[1]? s=None, int[1] dim=[-2,-1], str? norm=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:10138]: Nice following of SciPy's hfft2 signature

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -10135,6 +10135,26 @@
...
+  use_const_ref_for_mutable_tensors: True
+  python_module: fft
+  variants: function
+
+- func: fft_hfft2.out(Tensor self, int[1]? s=None, int[1] dim=[-2,-1], str? norm=None, *, Tensor(a!) out) -> Tensor(a!)
+  use_const_ref_for_mutable_tensors: True
+  python_module: fft
+  variants: function
+
+- func: fft_ihfft2(Tensor self, int[1]? s=None, int[1] dim=[-2,-1], str? norm=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:10148]: ihfft2 signature looks great, too

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -10167,6 +10187,26 @@
   python_module: fft
   variants: function
 
+- func: fft_hfftn(Tensor self, int[1]? s=None, int[1]? dim=None, str? norm=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:10190]: hfftn signature looks great

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -10167,6 +10187,26 @@
...
+  use_const_ref_for_mutable_tensors: True
+  python_module: fft
+  variants: function
+
+- func: fft_hfftn.out(Tensor self, int[1]? s=None, int[1]? dim=None, str? norm=None, *, Tensor(a!) out) -> Tensor(a!)
+  use_const_ref_for_mutable_tensors: True
+  python_module: fft
+  variants: function
+
+- func: fft_ihfftn(Tensor self, int[1]? s=None, int[1]? dim=None, str? norm=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:10200]: Awesome

File: test/test_spectral_ops.py [hunk 5]
```diff
@@ -25,6 +25,20 @@
 if TEST_LIBROSA:
     import librosa
 
+has_scipy_fft = False
```
Comment (mruberry) [inline test/test_spectral_ops.py:28]: Cool

File: test/test_spectral_ops.py [hunk 6]
```diff
@@ -25,6 +25,20 @@
...
+try:
+    import scipy.fft
+    has_scipy_fft = True
+except ModuleNotFoundError:
+    pass
+
+LooseVersion = distutils.version.LooseVersion
```
Comment (mruberry) [inline test/test_spectral_ops.py:35]: Nice versioning

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -6272,6 +6278,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                      dtypes=all_types_and_complex_and(torch.bool),
                      default_test_dtypes=floating_and_complex_types(),
                      check_batched_gradgrad=False),
+    SpectralFuncInfo('fft.hfftn',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6281]: For completeness we should add hfft2 and ihfft2, too (this can be done in a follow-up PR)

File: torch/fft/__init__.py [hunk 8]
```diff
@@ -808,6 +808,269 @@
...
+
+    >>> two_ffts = torch.fft.ifft(torch.fft.ihfft(t, dim=1), dim=0)
+    >>> torch.allclose(t, two_ffts)
+
+""".format(**common_args))
+
+hfftn = _add_docstr(_fft.fft_hfftn, r"""
+hfftn(input, s=None, dim=None, norm=None, *, out=None) -> Tensor
+
+Computes the n-dimensional discrete Fourier transform of a Herimitian symmetric
```
Comment (mruberry) [inline torch/fft/__init__.py:931]: Great first sentence

File: torch/fft/__init__.py [hunk 9]
```diff
@@ -808,6 +808,269 @@
...
+:attr:`input` signal.
+
+:attr:`input` is interpreted as a one-sided Hermitian signal in the time
```
Comment (mruberry) [inline torch/fft/__init__.py:934]: Cool

File: torch/fft/__init__.py [hunk 10]
```diff
@@ -808,6 +808,269 @@
...
+domain. By the Hermitian property, the Fourier transform will be real-valued.
+
+Note:
+    :func:`~torch.fft.hfftn`/:func:`~torch.fft.ihfftn` are analogous to
+    :func:`~torch.fft.rfftn`/:func:`~torch.fft.irfftn`. The real FFT expects
+    a real signal in the time-domain and gives Hermitian symmetry in the
+    frequency-domain. The Hermitian FFT is the opposite; Hermitian symmetric in
+    the time-domain and real-valued in the frequency-domain. For this reason,
+    special care needs to be taken with the shape argument :attr:`s`, in the
```
Comment (mruberry) [inline torch/fft/__init__.py:943]: This last sentence is a little vague but good enough for now

File: torch/fft/__init__.py [hunk 11]
```diff
@@ -808,6 +808,269 @@
...
+    property. In these cases the imaginary component will be ignored.
+    For example, any imaginary component in the zero-frequency term cannot
+    be represented in a real output and so will always be ignored.
+
+Note:
+    The correct interpretation of the Hermitian input depends on the length of
+    the original data, as given by :attr:`s`. This is because each input shape
+    could correspond to either an odd or even length signal. By default, the
+    signal is assumed to be even length and odd signals will not round-trip
+    properly. So, it is recommended to always pass the signal shape :attr:`s`.
```
Comment (mruberry) [inline torch/fft/__init__.py:957]: "So," can be removed and the last sentence can start with "It is..."

File: torch/fft/__init__.py [hunk 12]
```diff
@@ -808,6 +808,269 @@
...
+
+Args:
+    input (Tensor): the input tensor
+    s (Tuple[int], optional): Signal size in the transformed dimensions.
+        If given, each dimension ``dim[i]`` will either be zero-padded or
+        trimmed to the length ``s[i]`` before computing the real FFT.
+        If a length ``-1`` is specified, no padding is done in that dimension.
+        Defaults to even output in the last dimension:
+        ``s[-1] = 2*(input.size(dim[-1]) - 1)``.
```
Comment (mruberry) [inline torch/fft/__init__.py:966]: Should this be `2*input.size(dim[-1]) - 1` or is subtracting two correct?

File: torch/fft/__init__.py [hunk 13]
```diff
@@ -808,6 +808,269 @@
...
+    dim (Tuple[int], optional): Dimensions to be transformed.
+        The last dimension must be the half-Hermitian compressed dimension.
+        Default: all dimensions, or the last ``len(s)`` dimensions if :attr:`s` is given.
+    norm (str, optional): Normalization mode. For the forward transform
```
Comment (mruberry) [inline torch/fft/__init__.py:970]: Great norm description

File: torch/fft/__init__.py [hunk 14]
```diff
@@ -808,6 +808,269 @@
...
+        the exact inverse.
+
+        Default is ``"backward"`` (normalize by ``1/n``).
+
+Keyword args:
+    {out}
+
+Example:
+
+    >>> t = torch.rand(10, 10)
```
Comment (mruberry) [inline torch/fft/__init__.py:1053]: This example is cool but doesn't actually call ihfftn anywhere

File: torch/fft/__init__.py [hunk 15]
```diff
@@ -808,6 +808,269 @@
...
+    torch.Size([10, 9])
+    >>> torch.allclose(roundtrip, T)
+    True
+
+""".format(**common_args))
+
+ihfft2 = _add_docstr(_fft.fft_ihfft2, r"""
+ihfft2(input, s=None, dim=(-2, -1), norm=None, *, out=None) -> Tensor
+
+Computes the N-dimensional inverse discrete Fourier transform of real
```
Comment (mruberry) [inline torch/fft/__init__.py:874]: Great description

File: torch/fft/__init__.py [hunk 16]
```diff
@@ -808,6 +808,269 @@
...
+
+Example:
+
+    >>> T = torch.rand(10, 10)
+    >>> t = torch.fft.ihfft2(t)
+    >>> t.size()
+    torch.Size([10, 6])
+
+    Compared against the full output from :func:`~torch.fft.ifft2`, the
+    hermitian time-space signal takes up only half the space.
```
Comment (mruberry) [inline torch/fft/__init__.py:913]: Hermitian

File: torch/fft/__init__.py [hunk 17]
```diff
@@ -808,6 +808,269 @@
...
+    >>> torch.allclose(fftn[..., :6], rfftn)
+    True
+
+    The discrete Fourier transform is separable, so :func:`~torch.fft.ihfft2`
+    here is equivalent to a combination of :func:`~torch.fft.ifft` and
+    :func:`~torch.fft.ihfft`:
+
+    >>> two_ffts = torch.fft.ifft(torch.fft.ihfft(t, dim=1), dim=0)
+    >>> torch.allclose(t, two_ffts)
+
```
Comment (mruberry) [inline torch/fft/__init__.py:925]: True

**Review by mruberry — COMMENTED** [review]
Hey @peterbell10! This looks great overall. I made a few doc comments for your review. Nothing major but I think one example is still a placeholder. Just ping me when this is ready to go!

File: torch/fft/__init__.py (hunk 12 above)
Comment (peterbell10) [reply torch/fft/__init__.py:966]: Yes, subtracting two is correct because we are assuming an even output length. For a full signal of length `n`, the half-hermitian data layout has length `k == n // 2 + 1`. So, `n == (k - 1) * 2` for even `n`.

I like to think of this in terms of real degrees if freedom. The real signal has `n` degrees of freedom, but the complex signal doesn't have `2k`degrees of freedom because of the Hermitian property. For even `n`, two of the imaginary components are equal to their own conjugate image and so are always zero. That means there are `2k - 2` real degrees of freedom.

**Comment by peterbell10** [comment]
@mruberry PTAL

File: aten/src/ATen/native/SpectralOps.cpp [hunk 18]
```diff
@@ -416,32 +437,40 @@ Tensor& fft_rfftn_out(const Tensor& self,
...
     if (!s.has_value() || (s->back() == -1)) {
       const auto last_dim = desc.dim.back();
       return 2 * (self.sizes()[last_dim] - 1);
     }
     return desc.shape.back();
   }();
-  desc.shape.back() = last_dim_size / 2 + 1;
+  auto ld = last_dim_size / 2 + 1;
+  desc.shape.back() = ld;
+  TORCH_CHECK(ld >= 1, "Invalid number of data points (", last_dim_size, ") specified");
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:458]: `fname` should probably sneak in here, too

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @peterbell10!

Did you see that JAX added SciPy's [dct](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.fft.dct.html#jax.scipy.fft.dct) and  [dctn](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.fft.dctn.html?highlight=scipy.fft)? Might be time to revisit https://github.com/pytorch/pytorch/issues/49016. 

Looking at uses within Facebook, I can there are some uses of dct and dctn internally, and these ops seem reasonably popular on Github, too.

File: test/test_spectral_ops.py [hunk 19]
```diff
@@ -459,34 +452,106 @@ def test_fftn_invalid(self, device, dtype, op):
...
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float, torch.double)
+    def test_hfftn(self, device, dtype):
+        # input_ndim, dim
+        transform_desc = [
+            *product(range(2, 5), (None, (0,), (0, -1))),
+            *product(range(2, 5), (None,)),
+            (6, None),
+            (5, (1, 3, 4)),
+            (3, (0, -1)),
```
Comment (ngimel) [inline test/test_spectral_ops.py:465]: same here

File: torch/fft/__init__.py [hunk 20]
```diff
@@ -808,6 +808,271 @@
             -0.5000+0.6882j])
 """.format(**common_args))
 
+hfft2 = _add_docstr(_fft.fft_hfft2, r"""
+hfft2(input, s=None, dim=(-2, -1), norm=None, *, out=None) -> Tensor
+
+Computes the 2-dimensional discrete Fourier transform of a Herimitian symmetric
```
Comment (ngimel) [inline torch/fft/__init__.py:814]: typo: Herimitian -> Hermitian

File: torch/fft/__init__.py [hunk 21]
```diff
@@ -808,6 +808,271 @@
...
+:attr:`input` signal. Equivalent to :func:`~torch.fft.hfftn` but only
+transforms the last two dimensions by default.
+
```
Comment (ngimel) [inline torch/fft/__init__.py:817]: I think the note from `hfftn` that `input` is interpreted as a one-sided Hermitian signal in the time domain should also be repeated here, otherwise it's unclear what if `input` itself is required to have some symmetric properties.

**Comment by mruberry** [comment]
@peterbell10 just ping me when @ngimel's comments are addressed (thanks for taking a look, @ngimel!)

**Comment by peterbell10** [comment]
@mruberry PTAL.

**Comment by peterbell10** [comment]
@mruberry any update on this?

**Comment by mruberry** [comment]
It's just been hitting some internal unrelated land failures. I'm jedi landing it now to bypass them
